### ISSUE TITLE
TransferObserver.getBytesTransferred() always returns 0

### ISSUE THREAD

**Issue by Chronos2011** [aws/aws-sdk-android#217]
For a running upload with TransferUtility.upload(), the returned TransferObserver does always return 0 on TransferObserver.getBytesTransferred().

This happens even when uploading 16MB of data (which takes ~10 seconds).

**Comment by cristinaITdeveloper** [comment]
I have the same problem!
The upload is not working for any size of video.
The images upload works.

**Comment by wdane** [comment]
Are you calling Refresh or setting a TransferListener? Line 102 https://github.com/aws/aws-sdk-android/blob/master/aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/TransferObserver.java

If not, can you provide a code example?

**Comment by Chronos2011** [comment]
Yes, I am calling TransferObserver.refresh() immediately before calling TransferObserver.getBytesTransferred().

**Comment by Chronos2011** [comment]
Roughly, my code is as follows (I am converting the call from asynchronous to synchronous):

```
        AmazonS3 s3 = new AmazonS3Client(new BasicAWSCredentials(accessKeyId, secretAccessKey));
        s3.setRegion(Region.getRegion(Regions.fromName(region)));
        TransferUtility transferUtility = new TransferUtility(s3, context);
        File resultFile = new File(filename);
        TransferObserver transferObserver = transferUtility.upload(bucket, key, resultFile);
        loop:
        while (true) {
            transferObserver.refresh();
            TransferState state = transferObserver.getState();
            switch (state) {
                case COMPLETED:
                    break loop;
                case FAILED:
                    throw new TaskException(TaskExceptionType.UPLOAD_FAILED);
                default:
                    if (taskListener != null)
                        taskListener.onTaskProgress(this, new TaskProgress(transferObserver.getBytesTransferred(), transferObserver.getBytesTotal()));
            }
            try {
                Thread.sleep(100);
            } catch (InterruptedException interruptedException) {
                throw new TaskException(TaskExceptionType.UPLOAD_FAILED, interruptedException);
            }
        }

```

**Comment by karthiksaligrama** [comment]
What version of SDK are you using? And can you also tell me if this is on device or simulator? And if this is limited to version of Android?

**Comment by Chronos2011** [comment]
Using a Samsung GT-N7100 (Note II) running Android 4.4.2, API level 19. I don't have other devices handy at the moment - but I will check with my client whether I can test on other devices as well.

**Comment by Chronos2011** [comment]
No. I'm sorry, but it seems that nobody wants to pay for this.

**Comment by karthiksaligrama** [comment]
On a side note i'm note sure why you have it on a while loop. You can get around the problem by doing something like this

```
final TransferObserver observer = transferUtility.upload(bucket, key, resultFile);
observer.setTransferListener(new TransferListener() {
           @Override
            public void onStateChanged(final int id, final TransferState state) {
                
            }

            @Override
            public void onProgressChanged(final int id, final long bytesCurrent,
                                          final long bytesTotal) {
                
            }

            @Override
            public void onError(final int id, final Exception ex) {
                
            }
});
```

**Comment by Chronos2011** [comment]
I need a synchronous interface - and yes, I could synchronize the asynchronous interface with a latch, if I was sufficiently convinced that this solves my issue.
From your comment I understand that you could reproduce my issue (otherwise you couldn't be sure on a work-around). Can you provide a patch?

**Comment by karthiksaligrama** [comment]
yes it should be fixed in the next release.

**Comment by karthiksaligrama** [comment]
fixed with 2.3.8
